### DISCUSSION TITLE
Concurrency issues handling event streams

### DISCUSSION THREAD

**Discussion by poostwoud** [ravendb/ravendb#13182] [Q&A]
Hi,

I'm using Raven as a projection database as part of a CQRS model and experiencing concurrency issues handling event streams.

Every event handled undergoes the same ordinary flow: Get -> If exists, update, otherwise create. When trying to get the document the ChangeVector is stored in memory with the document. When updating, the document's ChangeVector is passed along with the Store method. 

However when multiple subsequent events are received to be handled, most events after the first event throw concurrency exceptions causing the events to be retried n-times or eventually be moved to the deadletter queue. I've already implemented optimistic concurrency, but this doesn't seem to be enough to prevent the concurrency exceptions. 

The projection is part of a microservice that must be scalable. Currently the concurrency issues are experienced with a single service as well as multiple services. So any solution must be able to span multiple instances of the same service as well.

Any thoughts on how to solve this? 

* Could it be a caching issue?
* Pessimistic locking? (https://ravendb.net/articles/pessimistic-locking-ravendb)
* Prevent "dirty" read (caching?) when using ACID transaction?
* ?

Cheers!

**Comment by grisha-kotler** [comment]
Hi Piter,

Are you updating the ChangeVector (that you store in memory) after you get the concurrency exception?

**Reply by poostwoud** [reply]
Hi,

Thanks for your reply. No I don't. The way of work is:

Get (cache ChangeVector) -> Apply event -> Update.

Below you'll find the methods used:

```csharp
        public OperationResult<TEntity> Get(TKey id)
        {
            try
            {
                using var session = _store.OpenSession();
                var document = session.Load<TEntity>(id.ToString());
                if (document == null) throw new IdentityNotFoundException<TKey>(id, _entityName);
                document.ChangeVector = session.Advanced.GetChangeVectorFor(document);
                return OperationResult<TEntity>.CreateSuccessResult(document);
            }
            catch (Exception ex)
            {
                return OperationResult<TEntity>.CreateFailure(ex);
            }
        }
        
        public OperationResult Update(TEntity entity)
        {
            try
            {
                //*****
                using var session = _store.OpenSession();

                //*****
                session.Advanced.UseOptimisticConcurrency = _store.UseOptimisticConcurrency;
                session.Store(entity, entity.ChangeVector, entity.Id.ToString());
                session.SaveChanges();

                //*****
                return OperationResult.CreateSuccessResult();
            }
            catch (Exception ex)
            {
                return OperationResult.CreateFailure(ex);
            }
        }        
```

Handling the event stops after the update has failed. So no update of the ChangeVector takes place.

Best regards,

Peter

**Reply by ayende** [reply]
Is this something that you can reproduce? What does the traffic watch says when you run this?

**Comment by aventus13** [comment]
Just curious- why do you even attempt to re-assign the change vector here?

```csharp
document.ChangeVector = session.Advanced.GetChangeVectorFor(document);
```

If it's the same document that you're updating, there should be no need for it. What happens if you don't do this? 
Also, if you want to avoid getting concurrency exception, you **shouldn't** be enabling optimistic concurrency.

**Reply by poostwoud** [reply]
Thanks for replying! I believe requesting the ChangeVector after loading the document is common practice? When I don't do this the ChangeVector property is empty. 

My problem is actually not the ConcurrencyException being thrown, but the fact that there is a rather big large delay on writing causing the getter to load a document that should be already modified in a previous store call, but still returns the old data.

**Reply by aventus13** [reply]
If the document (object) that you pass to the `Update` method is exactly the same object instance as the one you loaded from the database in the `Get` method, then there should be no need for manually assigning and passing in the change vector upon saving  the document.

As for the other part of your question, I'm afraid that I don't understand the problem enough to give you an accurate answer. Is the delay caused by the save operations (i.e. the delay is originating from the database) or by "delay" do you mean that you load the document, then some long-running processing happens and only then you attempt to save the document?

**Reply by poostwoud** [reply]
I've used example 4 as starting point: https://github.com/ravendb/docs/blob/eb476d52d360178e32bd9d2cb7742c3da5b0fb4a/Documentation/3.5/Samples/csharp/Raven.Documentation.Samples/ClientApi/Concurrency/UsingOptimisticConcurrencyInRealWorldScenarios.cs

and switched from Etag to ChangeVector.

**Reply by ayende** [reply]
What do you mean, large delay? Can you show a reproduction of this?

**Reply by ayende** [reply]
Note that when you read the change vector from the session, you get the value of the document's change vector _at the time it was loaded_.

**Reply by poostwoud** [reply]
Hi ayende, thanks for your feedback. For now I lost too much time on the matter, so I have switched to last write wins strategy by using a secondary source allowing me to offer a full document on every event, instead of applying the specific event. Not ideal, but it allowed me to move on in my project for now. I've tried to reproduce the scenario in parallel in a separate project, however (un)fortunately is doesn't occur anymore. There have been some changes in our configuration as well, so perhaps this did the job somehow... Thanks so far! I will close the discussion for now.
